Added the "insert" command and associated functionality and tests.

Tyler-g-hudson opened the pull request:
This command acquires a directory from the local machine and places it and its contents into a subfolder of the root folder on the given base image.

For this command I have added a CLI command, a supporting function to abstract that command, a dockerfile building function, and two tests intended to assess the validity of the dockerfile and also test the generated image.

I have also added some supporting infrastructure, a new pytest mark for build commands, and made some minor corrections that were missed in prior code.

Tyler-g-hudson committed:
Added the "insert" command and associated functionality and tests.

Tyler-g-hudson committed:
Removed an unused line from test_docker_insert

Tyler-g-hudson committed:
Renamed the default directory copy image.

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -55,12 +56,29 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
     add_tag_argument(parser=archive_parser, default="repo")
 
+    insert_dir_parser = subparsers.add_parser(
+        "insert",
```
I'm not sure about the name "insert" for this command. It seems too generic and non-descriptive.

WDYT about "copydir" instead?

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -55,12 +56,29 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+        type=Path,
+        required=True,
+        help="The directory to be copied to the image.",
+    )
```
Should we also give the user the option to specify the target directory (where the directory gets copied to in the container filesystem)?

gmgunter replied:
On a related note, it looks like this interface currently doesn't document where the directory gets copied to. I think that seems like important information that you'd want to convey to the user.

gmgunter commented on docker_cli/commands.py:
```diff
@@ -66,6 +67,61 @@ def get_archive(
+    of the given path. e.g. giving path "/tmp/dir/subdir" will result in the contents of
+    this path being saved in "/subdir" on the generated image.
+
+    This Dockerfile also places the working directory of the image inside of the copied
```
```suggestion
    This Dockerfile also changes the working directory of the image to the copied
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -66,6 +67,61 @@ def get_archive(
+    prefix = universal_tag_prefix()
+    img_tag = tag if tag.startswith(prefix) else f"{prefix}-{tag}"
+
+    dir_str = str(directory)
```
```suggestion
    dir_str = os.fspath(directory)
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -66,6 +67,61 @@ def get_archive(
+    # The absolute path of the given directory will be the build context.
+    # This is necessary because otherwise docker may be unable to find the directory if
+    # the build context is at the current working directory.
+    path_absolute = os.path.abspath(str(dir_str))
```
`dir_str` is already a string

```suggestion
    path_absolute = os.path.abspath(dir_str)
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -66,6 +67,61 @@ def get_archive(
+    if os.path.isdir(dir_str):
+        target_dir = os.path.basename(path_absolute)
+    else:
+        target_dir = os.path.basename(os.path.dirname(path_absolute))
```
Let's change the `else` branch here to raise an exception instead.

The input `directory` argument of this function is explicitly intended to be an existing directory on the host. If the user passed in something that wasn't a valid directory, we should raise an exception here to let them know about their mistake.

gmgunter commented on docker_cli/commands.py:
```diff
@@ -66,6 +67,61 @@ def get_archive(
+
+    # Build the image with the context at the absolute path of the given path.
+    return Image.build(
+        tag=img_tag, context=path_absolute, dockerfile_string=dockerfile, no_cache=True
```
>`no_cache=True`

Why are we disabling the cache here?

Tyler-g-hudson replied:
We had a discussion some time ago in which you mentioned that the build steps would all be run with the cache disabled. I interpreted that to include this step - at that time, did you mean that the cache would only be disabled for the steps following this one?

gmgunter replied:
Oh yeah, thanks for the reminder, you're right -- I do remember thinking that we might need to disable the cache for these build steps.

I think ultimately the desired behavior is this: if a user makes an update to the source code in their local repository on the host and then re-runs the build steps via wigwam, then Docker should correctly recognize that the code has changed and needs to be recompiled (rather than using a previously cached image, even though the Dockerfile didn't change).

But, in hindsight, Docker already handles that case correctly even if we don't manually disable the cache. As long as the user re-runs the ~`insert`~ `copydir` command after changing the source code, Docker will recognize that the contents of the directory have changed and will invalidate the cache.

So I can't quite remember exactly why I thought that we wanted to disable the cache. Maybe I wanted to ensure that the CI system would always rebuild the code on each PR update rather than using a cached image? But that would be the case anyway unless the PR was empty somehow (with no source code changes). Maybe I just didn't think it through properly or else had a different mental model for how this would work than what we ended up settling on :shrug:

I guess I would suggest removing the `no_cache=True` here unless you have a better memory than me about why we were thinking about disabling it. I can't think of any benefit to disabling it right now.

gmgunter commented on docker_cli/_docker_insert.py:
```diff
@@ -0,0 +1,46 @@
+    Returns a Dockerfile-formatted string which inserts a given directory on the
+    host machine into the image.
+
+    This Dockerfile also places the workdir into the inserted file.
```
```suggestion
    This Dockerfile also changes the working directory to the copied directory.
```

gmgunter commented on docker_cli/_docker_insert.py:
```diff
@@ -0,0 +1,46 @@
+    base : str
+        The base image tag.
+    target_dir : str | os.PathLike[str]
+        The position on the image to copy to.
```
```suggestion
        The path in the image filesystem to copy to.
```

gmgunter reviewed (commented):
Thanks! I think this PR size is basically perfect for me to get through easily.

The contents here looks great -- just had a few suggestions regarding the interface and docstring verbiage.

Tyler-g-hudson committed:
Added a custom target path to the  copydir command

~ insert command changed to "copydir"
~ Added the custom target path option to the command
~ Changed some wording and naming conventions, added some descriptive comments

Tyler-g-hudson committed:
Updated the comment on init_build_parsers to match the naming change

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -55,12 +56,38 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+        type=Path,
+        default=None,
+        help="The path on the image to copy the source directory to. If not given, "
+        "the lowest level directory of the path given by the directory argument "
```
Reading through the code, I had a hard time interpreting what you meant by "lowest level directory". I think "[base name](https://docs.python.org/3/library/os.path.html#os.path.basename)" might be the more commonly used term for this.

```suggestion
        "the base name of the path given by the directory argument "
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -66,6 +67,80 @@ def get_archive(
+    target_path : path-like or None
+        The directory to copy to, on the image, or None. If given, the contents of the
+        source directory will be copied to the given path. If None, the target path will
+        default to the lowest directory of the path given by the `directory` argument.
```
```suggestion
        default to the base name of the path given by the `directory` argument.
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -66,6 +67,80 @@ def get_archive(
+            target_dir = os.path.basename(path_absolute)
+        else:
+            raise ValueError(f"{dir_str} is not a valid directory on this machine.")
+            target_dir = os.path.basename(os.path.dirname(path_absolute))
```
This line is unreachable since it follows a `raise` statement. Let's remove it.

```suggestion
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -66,6 +67,80 @@ def get_archive(
+    # the machine, whereas a context at "." would be unable to see any directory that is
+    # not downstream of the working directory from which the program is called.
+    return Image.build(
+        tag=img_tag, context=path_absolute, dockerfile_string=dockerfile, no_cache=True
```
In our offline discussion, you suggested exposing the `no_cache` option to the user as a the command line option so they have more control over this and so it's easier for us to change our mind about what the best default behavior should be later on.

I think that's a good idea -- let's go ahead and expose that option here rather than hard-coding it to True.

I would suggest setting the default to False for now (see [this comment](https://github.com/isce3-testing/image-sandbox/pull/5/files#r1309388478) for context).

Tyler-g-hudson committed:
Added CLI no-cache option for get_archive and copy_dir

~ Also changed references to "lowest level directory" to instad say "base name"

gmgunter commented on docker_cli/cli/build_commands.py:
```diff
@@ -54,13 +55,48 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
 def build_command_names() -> List[str]:
     """Returns a list of all build command names."""
-    return ["get-archive"]
+    return ["get-archive", "insert"]
```
```suggestion
    return ["get-archive", "copydir"]
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -43,6 +45,8 @@ def get_archive(
     url_reader : URLReader | None, optional
         If given, will use the given URL reader to acquire the Git archive. If None,
         will check the base image and use whichever one it can find. Defaults to None.
+    no_cache : bool
```
```suggestion
    no_cache : bool, optional
```

gmgunter commented on docker_cli/commands.py:
```diff
@@ -63,7 +67,86 @@ def get_archive(
+        source directory will be copied to the given path. If None, the target path will
+        default to the base name of the path given by the `directory` argument.
+        Defaults to None.
+    no_cache : bool
```
```suggestion
    no_cache : bool, optional
```

gmgunter reviewed (approved):
LGTM thanks!

Tyler-g-hudson committed:
Last bits of documentation change
